### PR TITLE
Get notifications

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo/issues/308

### CODE REVIEW

File: neo/SmartContract/InteropService.cs [hunk 1]
```diff
@@ -201,6 +202,21 @@ private static bool Runtime_Serialize(ApplicationEngine engine)
...
 
+        private static bool Runtime_GetNotifications(ApplicationEngine engine)
+        {
+            var data = engine.CurrentContext.EvaluationStack.Pop().GetByteArray();
+            if (data.Length != UInt160.Length) return false;
+
+            var hash = new UInt160(data);
+            engine.CurrentContext.EvaluationStack.Push
+                (
+                new VM.Types.Array(engine.Notifications
```
Comment (erikzhang) [inline neo/SmartContract/InteropService.cs]: You should check the array size. Also, should we consider using `Enumerator`?

**Comment by shargon** [comment]
@igormcoelho we need to check if is a valid hash, and we don't have a constant for it

File: neo/SmartContract/InteropService.cs (hunk 1 above)
Comment (shargon) [reply neo/SmartContract/InteropService.cs]: What do you prefer, array or enumerator? i think that it could be easier if is an array?

**Comment by igormcoelho** [comment]
I mean, its faster to pass this part in another PR, in this one we focus on notifications only.. its faster.

File: neo/SmartContract/InteropService.cs (hunk 1 above)
Comment (igormcoelho) [reply neo/SmartContract/InteropService.cs]: Array is very limited, in size and resources...Enum gives us hope to filter data quickly, and use it efficiently even over array max limit.
A System.Enumerator.Filter would be nice too, the challenge is to pass a lambda inside it. I used to defend OPEVAL for that, but perhaps we can do something safer and simpler.

File: neo/SmartContract/InteropService.cs [hunk 2]
```diff
@@ -201,6 +202,22 @@ private static bool Runtime_Serialize(ApplicationEngine engine)
...
+        {
+            var data = engine.CurrentContext.EvaluationStack.Pop().GetByteArray();
+            if (data.Length != UInt160.Length) return false;
+            if (!engine.CheckArraySize(engine.Notifications.Count)) return false;
+
+            var hash = new UInt160(data);
+            engine.CurrentContext.EvaluationStack.Push
+                (
+                new VM.Types.Array(engine.Notifications
+                .Where(u => hash == UInt160.Zero || u.ScriptHash == hash)
```
Comment (erikzhang) [inline neo/SmartContract/InteropService.cs]: `hash == UInt160.Zero` What is this?

File: neo/SmartContract/InteropService.cs [hunk 3]
```diff
@@ -38,6 +38,7 @@ public static partial class InteropService
         public static readonly uint System_Runtime_Serialize = Register("System.Runtime.Serialize", Runtime_Serialize, 0_00100000, TriggerType.All);
         public static readonly uint System_Runtime_Deserialize = Register("System.Runtime.Deserialize", Runtime_Deserialize, 0_00500000, TriggerType.All);
         public static readonly uint System_Runtime_GetInvocationCounter = Register("System.Runtime.GetInvocationCounter", Runtime_GetInvocationCounter, 0_00000400, TriggerType.All);
+        public static readonly uint System_Runtime_GetNotifications = Register("System.Runtime.GetNotifications", Runtime_GetNotifications, 1000_00000000/*TODO*/, TriggerType.All);
```
Comment (erikzhang) [inline neo/SmartContract/InteropService.cs]:
```suggestion
        public static readonly uint System_Runtime_GetNotifications = Register("System.Runtime.GetNotifications", Runtime_GetNotifications, 0_00010000, TriggerType.All);
```

File: neo/SmartContract/InteropService.cs (hunk 2 above)
Comment (shargon) [reply neo/SmartContract/InteropService.cs]: Is for retrive all notifications, without filter

**Comment by shargon** [comment]
@lock9 I waited for the test until we decided if this is the correct address, now it's done :)

**Comment by vncoelho** [comment]
@erikzhang always with magical "optimize" at the end...ehuwhahahau

File: neo/SmartContract/InteropService.cs [hunk 4]
```diff
@@ -201,6 +202,21 @@ private static bool Runtime_Serialize(ApplicationEngine engine)
             return true;
         }
 
+        private static bool Runtime_GetNotifications(ApplicationEngine engine)
+        {
+            var data = engine.CurrentContext.EvaluationStack.Pop().GetByteArray();
+            if (data.Length != UInt160.Length) return false;
+            if (!engine.CheckArraySize(engine.Notifications.Count)) return false;
```
Comment (igormcoelho) [inline neo/SmartContract/InteropService.cs]: So we are limited to array size. Lets proceed and see how enums fit later.

**Comment by shargon** [comment]
@erikzhang if we want to know the hash of the notifications, is not possible, maybe is better to use Interop, and `GetNotificationHash` and `GetNotificationState`?

**Comment by erikzhang** [comment]
> if we want to know the hash of the notifications, is not possible

Return the `(hash, notification)` pair?

**Comment by shargon** [comment]
Please review it again, now is returned with the hash, so is possible to check which contract emit this notifications

**Comment by igormcoelho** [comment]
Congratulations Shargon! Important move for us!
I still prefer to create a specific InteropStackItem for that, with more capabilities in the future... returning two items simultaneously may overcomplicate syntax for C# devpack and other languages that don't support multiple returns.
Even C# uses the concept of `out` to keep its single return, plus others... so I think we will need to fix this. Anyway, congratulations again! Let's keep moving.

**Pull request merged**